### PR TITLE
*: add builtin aggregate function `var_pop`

### PR DESCRIPTION
**Pull request by githubFZX**
### What problem does this PR solve? 
[#7623](https://github.com/pingcap/tidb/issues/7623)

Tests
-------------------------------------------------------------------------------------------------------------------
```
mysql> create table numerical_t(
    -> a tinyint,
    -> b smallint,
    -> c mediumint,
    -> d int,
    -> e bigint,
    -> f float,
    -> g double,
    -> h decimal
    -> );
Query OK, 0 rows affected (0.10 sec)
```
-------------------------------------------------------------------------------------------------------------------
```
mysql> insert into numerical_t 
    -> values
    -> (1, 2, 3, 4, 5, 6.1, 7.2, 8.3), 
    -> (2, 3, 4, 5, 6, 7.2, 8.3, 9.4),
    -> (1, 3, 4, 5, 6, 7.1, 8.2, 9.3);
Query OK, 3 rows affected, 3 warnings (0.01 sec)
```
-------------------------------------------------------------------------------------------------------------------
```
mysql> select a, var_pop(b),var_pop(c),var_pop(d),var_pop(e),var_pop(f),var_pop(g),var_pop(h) from numerical_t group by a;
+------+------------+------------+------------+------------+------------+--------------------+------------+
| a    | var_pop(b) | var_pop(c) | var_pop(d) | var_pop(e) | var_pop(f) | var_pop(g)         | var_pop(h) |
+------+------------+------------+------------+------------+------------+--------------------+------------+
|    1 |       0.25 |       0.25 |       0.25 |       0.25 |       0.25 | 0.2500000000000071 |       0.25 |
|    2 |          0 |          0 |          0 |          0 |          0 |                  0 |          0 |
+------+------------+------------+------------+------------+------------+--------------------+------------+
2 rows in set (0.00 sec)

```

### CODE REVIEW

**Comment by zz-jason** [comment]
@githubFZX Thanks for your contribution, please follow the [Contribution Guide](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) to add a proper PR title.

**Comment by githubFZX** [comment]
Okay， thank you for your guidance and wish you have a happy life.




------------------ 原始邮件 ------------------
发件人: "Zhang Jian"<notifications@github.com>;
发送时间: 2019年7月17日(星期三) 下午5:35
收件人: "pingcap/tidb"<tidb@noreply.github.com>;
抄送: "1411679031"<1411679031@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [pingcap/tidb] Add var_pop (#11155)




@githubFZX Thanks for your contribution, please follow the Contribution Guide to add a proper PR title.

**Comment by zz-jason** [comment]
@githubFZX friendly ping, any update?

**Comment by zz-jason** [comment]
@githubFZX Hi, it has been a long time since the PR is filed. Any update to fix the CI failure?

File: executor/aggregate_test.go [hunk 1]
```diff
@@ -344,11 +352,11 @@ func (s *testSuite1) TestAggregation(c *C) {
 	_, err = tk.Exec("select std_samp(a) from t")
 	// TODO: Fix this error message.
 	c.Assert(errors.Cause(err).Error(), Equals, "[expression:1305]FUNCTION std_samp does not exist")
-	_, err = tk.Exec("select variance(a) from t")
+	//_, err = tk.Exec("select variance(a) from t")
+	//c.Assert(errors.Cause(err).Error(), Equals, "unsupported agg function: variance")
```
Comment (zz-jason) [inline executor/aggregate_test.go]: please do not comment them out, you can simply rewrite the expected behavior of this test.

File: executor/aggregate_test.go [hunk 2]
```diff
@@ -344,11 +352,11 @@ func (s *testSuite1) TestAggregation(c *C) {
...
 	// TODO: Fix this error message.
-	c.Assert(errors.Cause(err).Error(), Equals, "unsupported agg function: var_pop")
-	_, err = tk.Exec("select var_pop(a) from t")
-	c.Assert(errors.Cause(err).Error(), Equals, "unsupported agg function: var_pop")
+	//_, err = tk.Exec("select var_pop(a) from t")
+	//c.Assert(errors.Cause(err).Error(), Equals, "unsupported agg function: var_pop")
```
Comment (zz-jason) [inline executor/aggregate_test.go]: ditto

File: expression/aggregation/base_func.go [hunk 3]
```diff
@@ -233,6 +235,12 @@ func (a *baseFuncDesc) typeInfer4LeadLag(ctx sessionctx.Context) {
 	}
 }
 
+func (a *baseFuncDesc) typeInfer4VarPop(ctx sessionctx.Context) {
+	//var_pop's return value type is double
```
Comment (zz-jason) [inline expression/aggregation/base_func.go]:
```suggestion
	// var_pop's return value type is double
```

It's recommended to add a space between `//` and the first word.

**Comment by zz-jason** [comment]
@githubFZX please sign the CLA. Ref to https://github.com/pingcap/tidb/pull/11155#issuecomment-509660976

File: executor/aggfuncs/func_varpop.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+type varPop4Float64 struct {
+	baseAggFunc
+}
+
+type partialResult4Float64 struct {
```
Comment (zz-jason) [inline executor/aggfuncs/func_varpop.go]: It's better to use the algorithm described by Chan, Golub, and LeVeque in "Algorithms for computing the sample variance: analysis and recommendations" to calculate the variance. You can find how apache hive implement this at here: https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFVariance.java#L194

Comment (githubFZX) [reply executor/aggfuncs/func_varpop.go]: Okay. I will refer to it.

**Comment by zz-jason** [comment]
@githubFZX friendly ping, any update?

**Pull request closed without merging**